### PR TITLE
Add ability to calculate MD5 of Stream

### PR DESCRIPTION
**Pull request by sticilface**
Add ability to calculate MD5 of Stream

### CODE REVIEW

File: cores/esp8266/MD5Builder.cpp [hunk 1]
```diff
@@ -23,6 +23,47 @@ void MD5Builder::addHexString(const char * data){
   free(tmp);
 }
 
+bool MD5Builder::addStream(Stream & stream, const int total_len) {
+    const int buf_size = 512;
+    int bytesleft = total_len;
+    int bytestoread;
```
Comment (gtalusan) [inline cores/esp8266/MD5Builder.cpp]: bytestoread looks unused here.

File: cores/esp8266/MD5Builder.cpp [hunk 2]
```diff
@@ -23,6 +23,47 @@ void MD5Builder::addHexString(const char * data){
...
+                }
+
+                // not read more the buffer can handle
+                if(readBytes > buf_size) {
+                    readBytes = buf_size;
+                }
+
+                // read data  
+                int bytesread = stream.readBytes(buf, readBytes);
+                bytesleft -= bytesread;
```
Comment (gtalusan) [inline cores/esp8266/MD5Builder.cpp]: Also, here, I think bytesread should be subtracted after the check.. and handle the case where the read is short.

Comment (Links2004) [inline cores/esp8266/MD5Builder.cpp]: I think there can not be a short read since sizeAvailable is what we already have in the buffer to be used.

Comment (gtalusan) [inline cores/esp8266/MD5Builder.cpp]: That's assuming the implementation details of Stream, though.  If you imagine something like a GZIPStream that decompresses on read then you might have issues if there's a problem decompressing a block of the underlying data.

Comment (gtalusan) [inline cores/esp8266/MD5Builder.cpp]: But probably not a big deal if you're only interested in whether or not the MD5 matches.

Comment (Links2004) [inline cores/esp8266/MD5Builder.cpp]: yes in this case you are right, 
zip or gzip support on the ESP where nice, 
but haven't looked in to it yet.

**Comment by sticilface** [comment]
mmm is there a better way to do this, that doesn't involve me committing and uncommiting, and making a bit of a mess here.  I guess i can delete the pull request and just start afresh - that would loose comments though (i guess)... i do have a love hate with git.. any advice welcome?

**Comment by Links2004** [comment]
you can combine commits an make a forced push to your git.

**Comment by asetyde** [comment]
Hi , it isn’t possible do a staging version ? at now stable is unique

**Comment by sticilface** [comment]
mah.... i combined them... but it won't let me push as I'm technically behind!  ugh... hate git sometimes...

**Comment by sticilface** [comment]
Couldn't do it..  my git is not very hot.
see
https://github.com/esp8266/Arduino/pull/1344

**Pull request closed without merging**